cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs

```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Distribution.Solver.Types.WithConstraintSource
+  ( WithConstraintSource (..)
+  , showWithConstraintSource
+  , withUnknownConstraint
+  ) where
+
+import Distribution.Solver.Compat.Prelude
+
+import Distribution.Solver.Types.ConstraintSource (ConstraintSource (..), showConstraintSource)
+import Distribution.Pretty (Pretty (pretty))
+import Text.PrettyPrint
+
+-- | A package bundled with a `ConstraintSource`.
+data WithConstraintSource pkg =
```

Comment:
I see this type being used as `WithContraintSource UserConstraint` so I think `pkg` is not actually a `pkg`? i.e. `constraintPackage` is misnamed?

Reply:
I renamed them, how do these look?
- `constraintPackage` → `constraintInner`
- `constraintConstraint` → `constraintSource`